### ISSUE TITLE
[feat] Heatit Thermostat TF 056 (ThermoFloor AS) Hass support

### ISSUE THREAD

**Issue by magpern** [OpenZWave/Zwave2Mqtt#211]
I have a [Heatit floor thermostat](https://www.heatit.com/heating-control/floor-heating-thermostats/heatit-z-trm2fx/). How do I get it to show up in Home Assistant? I use autodiscovery, but wiki says climate device is not audodiscovery enabled.

In the documentation there is a section about custom device and the example shows a similar device [Heatit 021](https://github.com/OpenZWave/Zwave2Mqtt#custom-components) (mine is the 056).

I know the device worked fine in home assistant, then I had to reinstall the server and I cannot figure out how I got it working.
I don't understand the the documentation. What do I have to add, where? In home assistant? In zwave2mqtt? I don't have a store folder anywhere. I have a .store
I don't have a customDevices.json... If I create one, in the docker /store folder and past in the example, nothing happens.

If I run native zwave implementation in home assistant, it shows up, but incorrect. When I got it working some how before with zwave2mqtt, it showed up correctly.

Please if anyone could enlighten me, that would make my day!

**Comment by robertsLando** [comment]
@magpern If you are using docker the location of the store folder depends on how you start your docker container. If you are using volumes you can't easily access it, you should create a store folder and then map it inside the container.  

Check here for more info: https://github.com/robertsLando/Zwave2Mqtt-docker#run-using-local-folder
Or even better using docker-compose: https://github.com/robertsLando/Zwave2Mqtt-docker#run-as-a-service

**Comment by magpern** [comment]
@robertsLando I use the hass.io and the zwave2mqtt plug-in, so I assume everything is set up as it should.
This are my shares:
![image](https://user-images.githubusercontent.com/6589760/72875879-8426a280-3cf5-11ea-9ccd-27d69f91b65c.png)

But question remains, what files do I have to create to make the Heatit device work?

**Comment by robertsLando** [comment]
@magpern As documented in readme.md [here](https://github.com/OpenZWave/Zwave2Mqtt#custom-components)

> Starting from version 2.0.7 you can specify your custom devices configuration inside store/customDevices(.js|.json) file. This allows users that use Docker to create their custom hass devices configuration without the need to build a new container

So what you need to do is to create a file named `customDevices(.js|.json)` (extension js or json) in `store` folder of z2m (I don't know how the docker plugin maps that folder because I haven't created it). Maybe the configuration already existing for heatit works even for you: 

```js
{ // Heatit Thermostat TF 021 (ThermoFloor AS)
    type: 'climate',
    object_id: 'thermostat',
    values: ['64-1-0', '49-1-1', '67-1-1', '67-1-2'],
    mode_map: {'off': 'Off', 'heat': 'Heat (Default)', 'cool': 'Cool'},
    setpoint_topic: { "Heat (Default)": '67-1-1', "Cool": '67-1-2' },
    default_setpoint: '67-1-1',
    discovery_payload: {
      min_temp: 15,
      max_temp: 30,
      modes: ['off', 'heat', 'cool'],
      mode_state_topic: '64-1-0',
      mode_command_topic: true,
      current_temperature_topic: '49-1-1',
      current_temperature_template: '{{ value_json.value }}',
      temperature_state_template: '{{ value_json.value }}',
      temperature_command_topic: true
    }
  }
```

Check [here](https://github.com/OpenZWave/Zwave2Mqtt#thermostats) for more info about how to edit it

**Comment by magpern** [comment]
@robertsLando zwave2mqtt finds the device the "heatit z-trm2fx". It is shown in the web gui. But there is nothing in Home Assistant gui.
In documentation it says that climate devices will not be autodiscovered.

So, if I find this path "store" and put a file there called customeDevices.json and in this file put the text between the { } above, it will show up automatically in home assistant?

**Comment by magpern** [comment]
@robertsLando I removed the hassio add-on, and installed zwave2mqtt docker with docker compose instead.
No luck...

I see the device in zwave2mqtt with out-of.-the-box docker
![image](https://user-images.githubusercontent.com/6589760/72891912-9d3f4b80-3d15-11ea-8913-e49f29ce65b5.png)

I also see it in home assistant, but just as sensors
![image](https://user-images.githubusercontent.com/6589760/72892011-cc55bd00-3d15-11ea-938e-eccada8523e8.png)

I see no
![image](https://user-images.githubusercontent.com/6589760/72892087-f60ee400-3d15-11ea-95b3-298211636a78.png)

I tried to add the store/customDevices.json and copied in the text menchened above, but still no HVAC control in home assistant.

**Comment by robertsLando** [comment]
Wait wait you don't have to `just copy` what I give to you to make it working. If you read the documentation it is well explained that you need to specify the correct deviceid of your device in order to make it working... 

try to create `store/customDevices.js` with this inside:

```js
module.exports = {
  '411-514-3': [
    { // Heatit Thermostat TF 021 (ThermoFloor AS)
      type: 'climate',
      object_id: 'thermostat',
      values: ['64-1-0', '49-1-1', '67-1-1', '67-1-2'],
      mode_map: { 'off': 'Off', 'heat': 'Heat (Default)', 'cool': 'Cool' },
      setpoint_topic: { "Heat (Default)": '67-1-1', "Cool": '67-1-2' },
      default_setpoint: '67-1-1',
      discovery_payload: {
        min_temp: 15,
        max_temp: 30,
        modes: ['off', 'heat', 'cool'],
        mode_state_topic: '64-1-0',
        mode_command_topic: true,
        current_temperature_topic: '49-1-1',
        current_temperature_template: '{{ value_json.value }}',
        temperature_state_template: '{{ value_json.value }}',
        temperature_command_topic: true
      }
    }
  ]
}
```

**Comment by magpern** [comment]
@robertsLando Thanks! It is hard to follow a document that shows fragments of things when you don't understand what you have to do and why.

I changed the customeDevices.json to what you wrote above, I also tried to rename the file to *.js . I excluded the device from zwave network and included it again.

I see the device in MQTT under the homeassistant topic. It is registered as `homeassistant/sensor/nodeID_12`

In the home assistant gui all i see is only the small sensor circles in the top, no
![image](https://user-images.githubusercontent.com/6589760/72923511-83225f00-3d4f-11ea-9d91-3cf0ef6e03bf.png)

How do you get that into home assistant? Is there any trace files to look at?

**Comment by magpern** [comment]
@robertsLando I changed payload type to "JSON Time-Value" then the home assistant gui shows the 
![image](https://user-images.githubusercontent.com/6589760/72924450-3b043c00-3d51-11ea-9d2d-5d8c6d03acaa.png)
If I change it back to "Entire Zwave value Object" it is still there. Talk about confusing.
The "Heat" button is not working. Nothing happens and if I change the temp with the slider, the set temp is changed but it sends "Cool"
The payload type is not documented, impossible to know what it is used for. Sure, it says sends value time or complete zwave object, but why and how it is used is not described. That is very hard for the end user to understand

**Comment by robertsLando** [comment]
@magpern Sorry but what version of Z2M are you using? 

Z2M is born as a mqtt gateway and hass integration is born just after lot requests from hass users (I'm not an hass user), and as described in docs it is still a 'beta'. The hass integration should work with all payload types 

>  That is very hard for the end user to understand

It is not that hard if you know Mqtt protocol and hass. But If you think docs can be improved just tell me or send a PR

**Comment by robertsLando** [comment]
@magpern 

Edit the customDevices.json as follow:

```js
module.exports = {
  '411-514-3': [
    { // Heatit Thermostat TF 056 (ThermoFloor AS)
      type: 'climate',
      object_id: 'thermostat',
      values: ['64-1-0', '49-1-1', '67-1-1', '67-1-2'],
      mode_map: { 'off': 'Off', 'heat': 'Heat (Default)', 'cool': 'Cool' },
      setpoint_topic: { "Heat": '67-1-1', "Cool": '67-1-2' },
      default_setpoint: '67-1-1',
      discovery_payload: {
        min_temp: 15,
        max_temp: 30,
        modes: ['off', 'heat', 'cool'],
        mode_state_topic: '64-1-0',
        mode_command_topic: true,
        current_temperature_topic: '49-1-1',
        current_temperature_template: '{{ value_json.value }}',
        temperature_state_template: '{{ value_json.value }}',
        temperature_command_topic: true
      }
    }
  ]
}
```

**Comment by magpern** [comment]
I think it works now, with your latest customDevices.js.
I added `temp_step: 0.5` so slider allowes for 0.5 increments. Added my file content below.
I believe I run on the latest of everything, just installed it and docker downloads the latest of everything.

One small question remains. If I use the native zwave integration with home assistant, it will show "preset mode" as well. You are able to select "Energy Heat" or "None", as device has this mode. I think it is called `preset_modes`

In home assistant config files it looks something like this:
`                "capabilities": {
                    "hvac_modes": [
                        "heat",
                        "cool",
                        "off"
                    ],
                    "max_temp": 35,
                    "min_temp": 7,
                    "preset_modes": [
                        "Energy Heat",
                        "none"
                    ]
                },`

Is this also possible with the autodiscovery feature of zwave2mqtt?
Is it also possible to implement dimmer function if device supports it? The Heatit device can set the brightness of the LED display?

The customeDevices.js that now work:
`module.exports = {
  '411-514-3': [
    { // Heatit Thermostat TF 056 (ThermoFloor AS)
      type: 'climate',
      object_id: 'thermostat',
      values: ['64-1-0', '49-1-1', '67-1-1', '67-1-2'],
      mode_map: { 'off': 'Off', 'heat': 'Heat', 'cool': 'Cool' },
      setpoint_topic: { "Heat": '67-1-1', "Cool": '67-1-2' },
      default_setpoint: '67-1-1',
      discovery_payload: {
        min_temp: 15,
        max_temp: 30,
        temp_step: 0.5,    
        modes: ['off', 'heat', 'cool' ],
        mode_state_topic: '64-1-0',
        mode_command_topic: true,
        current_temperature_topic: '49-1-1',
        current_temperature_template: '{{ value_json.value }}',
        temperature_state_template: '{{ value_json.value }}',
        temperature_command_topic: true
      }
    }
  ]
}
`

**Comment by robertsLando** [comment]
@magpern Here you can find a list of parameters for MQTT HVACs: https://www.home-assistant.io/integrations/climate.mqtt/.

Just edit the discovery payload according to this propertiesv :)

**Comment by magpern** [comment]
@robertsLando Thanks! I have studied the MQTT HVAC documentation, but I guess I'm not smart enough.
I don't understand how the module.exports discovery_payload is transformed to a MQTT discovery message and then transformed in home assistant to a device entry.

eg. in module.exports there is a line `mode_command_topic: true`, in MQTT HVAC documentation it is supposed to be a string, in the MQTT server it has been represented as `"mode_command_topic":"zwave/nodeID_12/64/1/0/set"` Magic?

In MQTT HVAC the preset_mode is not mentioned at all, so adding it is just guessing, but adding it to module.exports gives errors in parsing, I have tried for hours.
preset_,mode is mentioned in [https://developers.home-assistant.io/docs/en/entity_climate.html](https://developers.home-assistant.io/docs/en/entity_climate.html) but how that correlates with MQTT HVAC and in the end module.exports is beyond me.

I would like to learn, but I just don't know where to start.

**Comment by robertsLando** [comment]
Climate entity is different from MQTT HVAC, the last one doesn't support all Climate features this is why you are getting that parse error, you can just use the options listed in the link I have provided.

About the `"mode_command_topic": true` that is there just to know that it needs to be calculated, as you can imagine that parameters is not equal for every thermostat but depends on node name/location and other gateway config. 

It is calculated [here](https://github.com/OpenZWave/Zwave2Mqtt/blob/master/lib/Gateway.js#L645)

**Comment by magpern** [comment]
@robertsLando ok, thanks! Maybe in the future I will get more involved in the understanding. But as for the current issue, I am good. The device works now in HA, and the extra features, like dimming of LED can be done through the web interface of zwave2mqtt. I will settle there.
